File: server/Services/KeepsService.cs
namespace keeper.Services;
public class KeepsService
{
  public KeepsService(KeepsRepository keepsRepository, VaultsService vaultsService, ProfilesService profilesService)
  {
    _keepsRepository = keepsRepository;
    _vaultsService = vaultsService;
    _profilesService = profilesService;
  }
  private readonly KeepsRepository _keepsRepository;
  private readonly VaultsService _vaultsService;
  private readonly ProfilesService _profilesService;


  internal Keep CreateKeep(Keep keepData)
  {
    Keep keep = _keepsRepository.CreateKeep(keepData);
    return keep;
  }

  internal List<Keep> GetAllKeeps()
  {
    List<Keep> keeps = _keepsRepository.GetAllKeeps();
    return keeps;
  }

  private Keep GetKeepById(int keepId)
  {
    Keep keep = _keepsRepository.GetKeepById(keepId);


    if (keep == null) throw new Exception($"Invalid keep id: {keepId}");
    return keep;
  }

  internal Keep IncrementViews(int keepId, string userId)
  {
    Keep keep = GetKeepById(keepId);

    if (keep.CreatorId != userId)
    {
      keep.Views++;
    }
    _keepsRepository.IncrementViews(keep);
    return keep;
  }

  internal Keep UpdateKeep(int keepId, string userId, Keep keepData)
  {
    Keep keep = GetKeepById(keepId);

    if (keep.CreatorId != userId) throw new Exception("YOU CANNOT EDIT ANOTHER USERS KEEP, BUDDDY");

    keep.Name = keepData.Name ?? keep.Name;
    keep.Description = keepData.Description ?? keep.Description;
    keep.Img = keepData.Img ?? keep.Img;

    _keepsRepository.UpdateKeep(keep);
    return keep;
  }

  internal string DeleteKeep(int keepId, string userId)
  {
    Keep keep = GetKeepById(keepId);

    if (keep.CreatorId != userId) throw new Exception("YOU CANNOT DELETE ANOTHER USERS KEEP, HOMIE");

    _keepsRepository.DeleteKeep(keepId);
    return $"Deleted {keep.Name}";
  }

  internal List<KeptKeep> GetKeepsByVaultId(int vaultId, string userId)
  {
    _vaultsService.GetVaultById(vaultId, userId);

    List<KeptKeep> keeps = _keepsRepository.GetKeepsByVaultId(vaultId);
    return keeps;
  }

  internal List<Keep> GetKeepsByProfileId(string profileId)
  {
    _profilesService.GetProfileById(profileId);

    List<Keep> keeps = _keepsRepository.GetKeepsByProfileId(profileId);
    return keeps;
  }
}

File: client/src/services/KeepsService.js
import { api } from "./AxiosService.js"
import { Keep } from "@/models/Keep.js"
import { AppState } from "@/AppState.js"
import { logger } from "@/utils/Logger.js"
import { KeptKeep } from "@/models/KeptKeep.js"

class KeepsService {

  async getKeepsByVaultId(vaultId) {
    AppState.keeps = []
    const response = await api.get(`api/vaults/${vaultId}/keeps`)
    AppState.keptKeeps = response.data.map(keep => new KeptKeep(keep))
  }
  async getMyKeeps(accountId) {
    const response = await api.get(`api/profiles/${accountId}/keeps`)
    const keeps = response.data.map(keep => new Keep(keep))
    AppState.keeps = keeps
  }
  async deleteKeep(keepId) {
    const response = await api.delete(`api/keeps/${keepId}`)
    const keepIndex = AppState.keeps.findIndex(keep => keep.id == keepId)
    AppState.keeps.splice(keepIndex, 1)
  }
  async createKeep(keepData) {
    const response = await api.post('api/keeps', keepData)
    const createdKeep = new Keep(response.data)
    AppState.keeps.unshift(createdKeep)
  }
  async GetKeepsByProfileId(profileId) {
    const response = await api.get(`api/profiles/${profileId}/keeps`)
    const keeps = response.data.map(keep => new Keep(keep))
    AppState.keeps = keeps
  }
  async getKeepById(keepId) {
    AppState.activeKeep = null
    const response = await api.get(`api/keeps/${keepId}`)
    const keep = new Keep(response.data)
    AppState.activeKeep = keep
  }
  async getAllKeeps() {
    AppState.keeps = []
    const response = await api.get('api/keeps')
    const keeps = response.data.map(keep => new Keep(keep))
    AppState.keeps = keeps
  }

}

export const keepsService = new KeepsService()